File: python/src/agent_framework/core/tool_registry.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ValidationError

from agent_framework.core.contracts import Tool, ToolExecutionContext, ToolRegistry
from agent_framework.core.domain import AgentError, AgentErrorCode, ToolCallRequest, ToolDefinition


class DefaultToolRegistry(ToolRegistry):
    def __init__(self) -> None:
        self._tools: dict[str, Tool] = {}

    def register(self, tool: Tool) -> None:
        self._tools[tool.definition.name] = tool

    def list(self, toolset: list[str] | None = None) -> list[ToolDefinition]:
        tools = [tool.definition for tool in self._tools.values()]
        if toolset:
            allowed = set(toolset)
            tools = [tool for tool in tools if tool.name in allowed]
        return sorted(tools, key=lambda item: item.name)

    def get(self, name: str) -> Tool | None:
        return self._tools.get(name)

    async def execute(self, ctx: ToolExecutionContext, call: ToolCallRequest) -> tuple[Any, str | None]:
        tool = self._tools.get(call.name)
        if tool is None:
            raise AgentError(f"Tool not found: {call.name}", AgentErrorCode.TOOL_NOT_FOUND)
        try:
            output = await tool.execute(ctx, call.arguments)
            return output, None
        except Exception as exc:  # noqa: BLE001
            return None, str(exc)


def create_pydantic_tool(
    tool_definition: ToolDefinition,
    schema: type[BaseModel],
    execute,
) -> Tool:
    class PydanticTool:
        definition = tool_definition

        async def execute(self, ctx: ToolExecutionContext, args: dict[str, Any]) -> Any:
             
            try:
                parsed = schema.model_validate(args)
            except ValidationError as exc:
                raise AgentError(str(exc), AgentErrorCode.VALIDATION_ERROR) from exc
            return await execute(parsed, ctx)

    return PydanticTool()

File: typescript/packages/core/src/tool-registry.ts
import { z } from 'zod';
import type { Tool, ToolCallRequest, ToolExecutionContext, ToolRegistry } from './types/contracts.js';
import { AgentError } from './types/contracts.js';

export class DefaultToolRegistry implements ToolRegistry {
  private tools = new Map<string, Tool>();

  register(tool: Tool): void {
    this.tools.set(tool.definition.name, tool);
  }

  unregister(name: string): void {
    this.tools.delete(name);
  }

  list(toolset?: string[]): import('./types/domain.js').ToolDefinition[] {
    const all = [...this.tools.values()].map((tool) => tool.definition);
    if (!toolset || toolset.length === 0) {
      return all.sort((a, b) => a.name.localeCompare(b.name));
    }
    const allowed = new Set(toolset);
    return all.filter((tool) => allowed.has(tool.name)).sort((a, b) => a.name.localeCompare(b.name));
  }

  get(name: string): Tool | undefined {
    return this.tools.get(name);
  }

  async execute(
    ctx: ToolExecutionContext,
    call: ToolCallRequest,
  ): Promise<{ output: unknown; error?: string }> {
    const tool = this.tools.get(call.name);
    if (!tool) {
      throw new AgentError(`Tool not found: ${call.name}`, 'TOOL_NOT_FOUND');
    }

    try {
      const output = await withTimeout(
        tool.execute(ctx, call.arguments),
        ctx.abortSignal,
      );
      return { output };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { output: null, error: message };
    }
  }
}

export function createZodTool<T extends z.ZodTypeAny>(
  definition: import('./types/domain.js').ToolDefinition,
  schema: T,
  execute: (ctx: ToolExecutionContext, args: z.infer<T>) => Promise<unknown>,
): Tool {
  return {
    definition,
    async execute(ctx, args) {
      const parsed = schema.safeParse(args);
      if (!parsed.success) {
        throw new AgentError(parsed.error.message, 'VALIDATION_ERROR');
      }
      return execute(ctx, parsed.data);
    },
  };
}

async function withTimeout<T>(promise: Promise<T>, abortSignal?: AbortSignal): Promise<T> {
  if (!abortSignal) {
    return promise;
  }
  if (abortSignal.aborted) {
    throw new AgentError('Operation cancelled', 'CANCELLED');
  }
  return new Promise<T>((resolve, reject) => {
    const onAbort = () => reject(new AgentError('Operation cancelled', 'CANCELLED'));
    abortSignal.addEventListener('abort', onAbort, { once: true });
    promise.then(
      (value) => {
        abortSignal.removeEventListener('abort', onAbort);
        resolve(value);
      },
      (error) => {
        abortSignal.removeEventListener('abort', onAbort);
        reject(error);
      },
    );
  });
}
